### PR TITLE
Support streaming output of runtime env setup to logger/driver

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
This PR adds loggers for runtime env setup function. The runtime setup hook that is supposed to be ran at the runtime env agent can call an api `from ray.workers.pluggable_runtime_env import get_hook_looger` to retrieve it's logger. 

The logger will write to a per-job log file, which will then be picked up by the log monitor.

Here's a sample output with test script
```python
import ray
import os

ray.init(runtime_env={"working_dir": ".", "pip": ["requests"]})

@ray.remote
def task():
    d = os.getcwd()
    return d

print(ray.get(task.remote()))
```

```
➜  /tmp/nested RAY_RUNTIME_ENV_LOCAL_DEV_MODE=1 python test.py                                                                                                                                    
2021-07-26 18:13:25,262	INFO services.py:1253 -- View the Ray dashboard at http://127.0.0.1:8265
2021-07-26 18:13:25,505	INFO runtime_env.py:563 -- gcs://_ray_pkg_5874a6d06cc3dc0ee491dcead52d7cb6.zip doesn't exist. Create new package with . and None
2021-07-26 18:13:25,507	INFO runtime_env.py:570 -- gcs://_ray_pkg_5874a6d06cc3dc0ee491dcead52d7cb6.zip has been pushed with 498 bytes
(pid=runtime_env) 2021-07-26 18:13:27,518	WARNING setup_runtime_env.py:258 -- Current Ray version could not be detected, most likely because you are using a version of Ray built from source.  If you wish to use runtime_env, you can try building a wheel and including the wheel explicitly as a pip dependency.
(pid=runtime_env) 2021-07-26 18:13:27,523	INFO setup_runtime_env.py:115 -- Setting up conda environment with {'_inject_current_ray': True, '_ray_commit': '{{RAY_COMMIT_SHA}}', 'conda': None, 'env_vars': None, 'pip': 'requests\n', 'uris': ['gcs://_ray_pkg_5874a6d06cc3dc0ee491dcead52d7cb6.zip'], 'working_dir': '.'}
(pid=runtime_env) 2021-07-26 18:13:28,530	INFO conda.py:106 -- Creating conda environment /tmp/ray/session_2021-07-26_18-13-23_558291_75634/runtime_resources/conda/ray-7045e7e48fb6e755a40c993d50b43b817d6689dd
(pid=runtime_env) 2021-07-26 18:13:46,320	WARNING setup_runtime_env.py:68 -- Injecting /Users/simonmo/Desktop/ray/ray/python to environment /tmp/ray/session_2021-07-26_18-13-23_558291_75634/runtime_resources/conda/ray-7045e7e48fb6e755a40c993d50b43b817d6689dd because _inject_current_ray flag is on.
(pid=runtime_env) 2021-07-26 18:13:46,320	INFO setup_runtime_env.py:138 -- Finished setting up runtime environment at /tmp/ray/session_2021-07-26_18-13-23_558291_75634/runtime_resources/conda/ray-7045e7e48fb6e755a40c993d50b43b817d6689dd
(raylet) /Users/simonmo/Desktop/ray/ray/python/ray/autoscaler/_private/cli_logger.py:63: FutureWarning: Not all Ray CLI dependencies were found. In Ray 1.4+, the Ray CLI, autoscaler, and dashboard will only be usable via `pip install 'ray[default]'`. Please update your install command.
(raylet)   "update your install command.", FutureWarning)
/private/tmp/ray/session_2021-07-26_18-13-23_558291_75634/runtime_resources/_ray_pkg_5874a6d06cc3dc0ee491dcead52d7cb6
```

Note that I removed few duplicated message because there are still too many workers started when runtime env is blocked.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
This PR is ready for review pending some tests.

**Review by architkulkarni — APPROVED** [review]
Looks great, just added a few nits.  My main question is will the logging also work for the Client Server setup case?  (For example, when calling `ray.client(...).env(...).connect()`).  I guess it should, since the setup in this case still goes through `setup_runtime_env()` which is where all the logging happens.

File: dashboard/modules/runtime_env/runtime_env_agent.py [hunk 1]
```diff
@@ -29,17 +32,35 @@ def __init__(self, dashboard_agent):
...
+            self._per_job_logger_cache[job_id] = per_job_logger
+        return self._per_job_logger_cache[job_id]
+
     async def CreateRuntimeEnv(self, request, context):
         async def _setup_runtime_env(serialized_runtime_env, session_dir):
+            # This function will be ran inside a thread
+            def run_setup_with_logger():
+                runtime_env: dict = json.loads(serialized_runtime_env or "{}")
+                per_job_logger = self.get_or_create_logger(request.job_id)
+                with using_thread_local_logger(per_job_logger):
```
Comment (edoakes) [inline dashboard/modules/runtime_env/runtime_env_agent.py:58]: why is this necessary? it's not obvious to me so maybe leave a comment

File: python/ray/_private/log_monitor.py [hunk 2]
```diff
@@ -146,6 +151,12 @@ def update_log_filenames(self):
                     job_id = None
                     worker_pid = None
 
+                if "runtime_env" in file_path:
```
[the remark this answers is not in the document]
Comment (simon-mo) [reply python/ray/_private/log_monitor.py:154]: ah this is because substring matching is faster than regex matching (70ns vs 250ns).
